File: Behavioral/Command/javascript/command.js
/**
 * Created by HP on 3/14/2018.
 */

class Shop {
    constructor (shopName){
        this.shopName = shopName;
        this.command = [];
    }

    buy(item) {
        console.log(`Purchasing ${item} in ${this.shopName}`)
    }

    takeALook(item) {
        console.log(`Checking out ${item} in ${this.shopName}`)
    }

    returnItem(item) {
        console.log(`Returning ${item} in ${this.shopName}`)
    }

    execute(command, item) {
        return this[command].call(this, item);
    }
}

const sevenEleven = new Shop('7/11');

sevenEleven.execute('takeALook', 'iced coffee');
sevenEleven.execute('buy', 'sneekers');
sevenEleven.execute('buy', 'shower gel');
sevenEleven.execute('returnItem', 'shower gel');

File: Behavioral/Command/kotlin/Command.kt
interface ShopCommand {
    fun execute();
}

class BuyCommand(val item: String) : ShopCommand {
    override fun execute() = println("Purchasing item: $item");
}

class TakeALookCommand(val item: String) : ShopCommand {
    override fun execute() = println("Checking out item: $item");
}

class ReturnItemCommand(val item: String) : ShopCommand {
    override fun execute() = println("Returning: $item");
}

class ShopCommandProcessor {
    private val queue = ArrayList<ShopCommand>();

    fun addToQueue(shopCommand: ShopCommand) : ShopCommandProcessor = apply {
        queue.add(shopCommand);
    }

    fun processCommands() : ShopCommandProcessor = apply {
        queue.forEach(it.execute);
        queue.clear();
    }
}

fun main(args: Array<String>) {
    ShopCommandProcessor.addToQueue(BuyCommand("Iced Coffe"))
            .addToQueue(TakeALookCommand("Sneekers"))
            .addToQueue(ReturnItemCommand("Shower Gel"))
            .processCommands();
}